src/bin/purge_ttl.rs

```diff
@@ -69,32 +76,192 @@ pub fn statsd_from_env() -> Result<StatsdClient, Box<dyn Error>> {
+}
+
+fn continue_transaction(
+    session: &Session,
+    transaction_id: Vec<u8>,
+) -> Result<ExecuteSqlRequest, Box<dyn Error>> {
+    let mut ts = TransactionSelector::new();
+    ts.set_id(transaction_id);
+    let mut req = ExecuteSqlRequest::new();
+    req.set_session(session.get_name().to_string());
+    req.set_transaction(ts);
     Ok(req)
 }
 
+fn commit_transaction(
+    client: &SpannerClient,
+    session: &Session,
+    txn: Vec<u8>,
+) -> Result<(), Box<dyn Error>> {
+    let mut req = CommitRequest::new();
+    req.set_session(session.get_name().to_owned());
+    req.set_transaction_id(txn);
+    client.commit(&req)?;
+    Ok(())
+}
+
+pub struct SyncResultSet {
+    result: googleapis_raw::spanner::v1::result_set::ResultSet,
+}
+
+impl Iterator for SyncResultSet {
+    type Item = Vec<Value>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let rows = &mut self.result.rows;
+        if rows.is_empty() {
+            None
+        } else {
+            let row = rows.remove(0);
+            Some(row.get_values().to_vec())
+        }
+    }
+}
+
+fn delete_incremental(
+    client: &SpannerClient,
+    session: &Session,
+    table: String,
+    column: String,
+    chunk_size: u64,
+) -> Result<(), Box<dyn Error>> {
+    let (mut req, mut txn) = begin_transaction(&client, &session, RequestType::ReadWrite)?;
+    let select_sql = format!("SELECT fxa_uid, fxa_kid, collection_id, {} FROM {} WHERE expiry < CURRENT_TIMESTAMP() LIMIT {}", column, table, chunk_size);
+    trace!("Selecting rows to delete: {}", select_sql);
+    req.set_sql(select_sql.clone());
+    let mut result = SyncResultSet {
+        result: client.execute_sql(&req)?,
+    };
+
+    let mut total: usize = 1;
+    while total < 1_000_000 {
+        if result.result.rows.is_empty() {
+            info!("{}: done", table);
+            break;
+        }
+        let mut delete_sql = format!(
+            "DELETE FROM {} WHERE (fxa_uid, fxa_kid, collection_id, {}) IN (",
+            table, column,
+        );
+        for row in &mut result {
+            // Count starting at 1 so that i % chunk_size is false when on the first row
+            let fxa_uid = row[0].get_string_value().to_owned();
+            let fxa_kid = row[1].get_string_value().to_owned();
+            let collection_id = row[2].get_string_value().parse::<i32>().unwrap();
+            let id = row[3].get_string_value().to_owned();
+            trace!(
+                "Selected row for delete: i={} fxa_uid={} fxa_kid={} collection_id={} {}={}",
+                total,
+                fxa_uid,
+                fxa_kid,
+                collection_id,
+                column,
+                id
+            );
+            delete_sql = format!(
+                "{}('{}', '{}', {}, '{}'), ",
+                delete_sql, fxa_uid, fxa_kid, collection_id, id
+            );
+
+            total += 1;
+        }
+        delete_sql = format!("{}('', '', 0, ''))", delete_sql);
```

Comment:
I'm probably just paranoid but it'd be nice to avoid this end case in the delete. Could keep an eye on the index while iterating to decide whether or not to omit a comma or not

Reply:
Another trick I used to use is to remove the trailing comma. I'll switch to that kind of impl.